### DISCUSSION TITLE
How to make docstring format work?

### DISCUSSION THREAD

**Discussion by s-banach** [astral-sh/ruff#9439] [Q&A]
I have the following lines in `ruff.toml`:

```
target-version = "py39"
line-length = 88
pycodestyle.max-doc-length = 100

[format]
line-ending = "lf"
preview = true
docstring-code-format = true
```

I've tried adding code in my docstring in two ways:

A:
```
    Examples
    --------
    >>> import my_package
    >>> do_stuff(...)
```

B:
```
    Examples
    --------
    .. code-block:: python

        import my_package
        do_stuff(...)
```

But the ruff extension never seems to format this code in the docstring.
What am I doing wrong?

**Comment by charliermarsh** [comment]
Are you using the Ruff CLI, or the Ruff extension in your editor?

**Reply by s-banach** [reply]
Ruff extension in VSCode.
Whether I choose `Ruff: Format Document` or `Ruff: Fix All Auto-Fixable Problems`, nothing seems to happen.

**Reply by s-banach** [reply]
The extension works for formatting and linting everything else.

**Comment by charliermarsh** [comment]
I believe the code block needs to be indented:

```python
def func():
    """Foo

    Examples
    --------

    .. code-block:: python
        import my_package
        do_stuff(...)
        x=1
    """
    ...
```

\cc @BurntSushi who knows the rules.

**Reply by BurntSushi** [reply]
Right. And you should need a new line after `.. code-block`. From the [rST spec](https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#directives):

> The directive block consists of any text on the first line of the directive after the directive marker, and any subsequent indented text.

and

> Arguments and options must form a contiguous block beginning on the first or second line of the directive; a blank line indicates the beginning of the directive content block.

**Reply by s-banach** [reply]
I fixed the code so it builds with sphinx, but the formatter still doesn't work for me.

**Reply by BurntSushi** [reply]
Can you provide a complete source example that isn't working for you _and_ is valid reST?

**Reply by s-banach** [reply]
I figured out that the issue is this character: `\`

```
class MyClass:
    """
    Class for doing something fun.

    * | Bullet one. This is a really really really really really\
        reallyreally really really really long line.
    * | Bullet two.

    Examples
    --------

    .. code-block:: python

        my_class = MyClass(xs=[1,2,3, 4, 5, 6, 7, 8, 9])
    """

    def __init__(self, xs: list) -> None:
        """
        Parameters
        ----------
        xs : list
            The list of x data to store in the class.
        """
        self.xs = xs

```

**Reply by BurntSushi** [reply]
Yeah it looks like this behavior is inherited from Black?

https://github.com/astral-sh/ruff/blob/ba71772d935cb04229e9a65c032468c6674437d5/crates/ruff_python_formatter/src/string/docstring.rs#L108-L111

I don't know the underlying motivation of it though. cc @MichaReiser

**Reply by MichaReiser** [reply]
The reason Black avoids normalizing docstrings containing string continuations in docstrings is that normalizing them leads to AST changes:

From the Black documentation

https://github.com/psf/black/blob/b1d17ef9a2579d10e5e31d3bb9e0a31dee95d7e9/src/black/linegen.py#L431-L443
